### ISSUE TITLE
Feedback from first course delivery of lesson 2

### ISSUE THREAD

**Issue by cc-a** [ImperialCollegeLondon/grad_school_software_engineering_course#11]
This feedback is based on delivery of lesson 2 of the course via the graduate school on 21st Feb. Most attendees knew Python and had been programming for >1 year. Most attendees were using their own laptops and had successfully completed the setup instructions.

# Instructor/helper feedback on session

* Solution to main exercise needs updating in line with change to suggested input.
* Several users did not realise they had to create a new environment for this
  part of the course.

# Technical Issues Encountered
* Some users had to restart VS Code for the testing side bar icon appear
* At least one user the testing side bar icon would not appear (environment was
  setup correctly).

# Positive Attendee feedback
* Good slides and links to other resources
* Good resources
* Good balance between theory and practice today
* Good links to resources for other programming languages
* Useful to have extensive example to work on
* Useful to see various testing implementations (e.g. pytest or just assert)
* The amount of helpers on hand is really good
* Helpers know what they are talking about + useful for setup
* Very helpful advise when I had questions
* Lots of help + resources well taught
* Good content just fast
* Good exercise provided
* Emphasised the importance of unit testing software in general
* Explaining things in simple terms

# Negative Attendee feedback
* Examples should be more clear
* Online instructions not really clear
* Sometimes it was too general and I'd have liked a deeper explanation
* The testing explanation was a bit dense, could have been simplified further
* Give output of step test beforehand
* I don't think this should be advertised as suitable for 'basic knowledge of
  any programming language'. I found it hard to follow most of it.
* Not really familiar with Python so spent half of my time trying to understand
  how to run things and understanding how the code works
* Too much content for 4 hours, I would say make it 6 hours or split into 2
  separate courses.
* Pace was a bit fast, session would be good for longer
* Feel a bit rushed/steep learning curve.  May be useful to provide some easier
  examples before the hear diffusion problem
* Perhaps an example not using numpy etc. People work on all sorts (myself,
  images for example)

# Suggestions
* As suggested by one attendee providing the expected output of the heat
  equation step function would allow attendees to focus on testing principles
  and get to the more advanced material. Could also just refactor the analytical
  solution function to be more readily reusable.
* Even some Python users had not seen list comprehensions before, the example
  code may be more readable if rewritten with numpy arrays especially for
  non-Python users.
* Could provide an alternative exercise working on a different type of problem
  to accommodate users with less physical science backgrounds.

Anything to add @mwoodbri, @jcohen02?

**Comment by mwoodbri** [comment]
Thanks @cc-a.

All seems fair to me... I think some people were overthinking the exercise. I've amended it to clarify what the test should be checking but will consider alternatives!

Can also remove some superfluous material. Leave it with me.

@jcohen02: Any other thoughts much appreciated.

**Comment by jcohen02** [comment]
Sorry to be a bit late with this but a couple of thoughts/points that I wanted to add - I think these are also reflected in the feedback to some extent...

### Technical issues
 - As with the first class, I noticed a few VS Code issues where supposedly correct and similar set ups look different on different platforms, toolbars wouldn't appear etc. This is highlighted above in the context of the testing side bar but I think the issues went further than this. Aside from adding more time to resolve problems, I'm not sure there's a straightforward solution to this if people are using a combination of their own machines and lab machines?

### General feedback/thoughts
 - I thought the pace of presenting the material didn't feel rushed in itself. However the issue I observed was that there were quite a few questions and even if a question took 2-3 minutes to resolve, the people had in some cases fallen behind by this point and catching up might be difficult.
 - Can we look to build in more in the way of gaps into the presentation of the material? I guess a challenge of this is that some people may then find the pace too slow. (and it will of course require a longer slot for the course)
 - I felt that the temperature example, in particular, caused some confusion and while it was a nice example, it might be good to see if a less technical example could be used to make the intended point?

**Comment by AdrianDAlessandro** [comment]
I ran through the heat equation exercise. I have a couple of observations.

- I did not see a "Run Test" button, although I was using system python and venv instead of Conda.
- It was quite tricky to work out how the `step()` function we are meant to write is related to eq (1) in the theory section.
  - Not immediately obvious what `u` is and what `i` is.
  - Perhaps the point that you are testing an exact solution against a numerical solution is missed because the `step()` function is actually a part of the numerical solution.
  - It took me 45 minutes for the first **Refactoring** section alone. I believe most of this time was spent translating from the code to the theory/equations.
- In the instability section, it says "Check that this test fails..." perhaps it is important to highlight that it should fail for the right reason. It will fail if they haven't imported the relevant part of pytest
- I would not expect most people to understand what instability is in this context
- Should the code just import pytest entirely? Otherwise you need to keep adding functions to the import statement as you add tests
